### PR TITLE
Feature #7 mikkelinyt

### LINKED ISSUE #7: MikkeliNyt integration
As a developer i would like to design MikkeliNyt integration

### PR DESCRIPTION
**Pull request by anttileppa**
Resolves #7

### CODE REVIEW

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/persistence/dao/SystemSettingDAO.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+   * @param value new value
+   * @return updated system setting
+   */
+  public SystemSetting updateValue(SystemSetting setting, String value) {
+    setting.setValue(value);
+    return persist(setting);
+  }
+  
+  @Override
+  public void delete(SystemSetting setting) {
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/persistence/dao/SystemSettingDAO.java]: MINOR Remove this method to simply inherit it. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1185)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/ptv/PtvClient.java [hunk 2]
```diff
@@ -13,32 +10,27 @@
...
 
 @Dependent
 public class PtvClient extends fi.otavanopisto.ptv.client.ApiClient {
 
   private static final String BASE_PATH = "http://ptvenv.cloudapp.net:1494";
-  private static final Logger logger = Logger.getLogger(PtvClient.class.getName());
-  
+
+  @Inject
+  private Logger logger;
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/ptv/PtvClient.java:34]: MINOR Remove the "logger" field and declare it as a local variable in the relevant methods. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1450)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/rest/OrganizationsApiImpl.java [hunk 3]
```diff
@@ -101,6 +116,88 @@ public Response listServiceClasses(String organizationIdParam) {
...
+      .build();
+  }
+
+  @Override
+  public Response listOrganizationEvents(String organizationIdParam) {
+    OrganizationId organizationId = new OrganizationId(KuntaApiConsts.IDENTIFIER_NAME, organizationIdParam);
+    
+    List<Event> result = new ArrayList<>();
+   
+    // TODO: Sort events
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/rest/OrganizationsApiImpl.java:208]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1135)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/rest/OrganizationsApiImpl.java [hunk 4]
```diff
@@ -127,5 +224,18 @@ public Response listServiceClasses(String organizationIdParam) {
     
     return Collections.unmodifiableList(result);
   }
+  
+  private List<EventProvider> getEventProviders() {
+    // TODO: Prioritize providers
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/rest/OrganizationsApiImpl.java:244]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1135)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/IdController.java [hunk 5]
```diff
@@ -8,12 +8,27 @@
...
+ * @author Antti Leppä
+ */
 @Dependent
 public class IdController {
   
+  private IdController() {
+  }
+  
   @Inject
   private Instance<IdProvider> idProviders;
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/IdController.java]: MINOR Move this variable to comply with Java Code Conventions. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1213)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+    result.setName(stripHtml(nytEvent.getName()));
+    result.setOriginalUrl(nytEvent.getUrl());
+    result.setPlace(stripHtml(nytEvent.getPlace()));
+    result.setZip(stripHtml(nytEvent.getZip()));
+    result.setStart(parseOffsetDateTime(nytEvent.getStart()));
+    
+    return result;
+  }
+  
+  private static String stripHtml(String html) {
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java]: MAJOR Remove this unused private "stripHtml" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+    return StringUtils.trim(StringUtils.normalizeSpace(StringEscapeUtils.unescapeHtml4(html.replaceAll("\\<.*?>"," "))));
+  }
+  
+  private static OffsetDateTime parseOffsetDateTime(String text) {
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java]: MAJOR Remove this unused private "parseOffsetDateTime" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+    if (kuntaApiId == null) {
+      logger.info(String.format("Found new MikkeliNyt attachment %s", imageId));
+      Identifier newIdentifier = identifierController.createIdentifier(IdType.ATTACHMENT, MikkeliNytConsts.IDENTIFIER_NAME, imageId);
+      kuntaApiId = new AttachmentId(KuntaApiConsts.IDENTIFIER_NAME, newIdentifier.getKuntaApiId());
+    }
+    
+    return kuntaApiId;
+  }
+
+  private List<Event> transform(List<fi.otavanopisto.mikkelinyt.model.Event> nytEvents) {
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java:353]: MAJOR Remove this unused private "transform" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+      } finally {
+        imageClient.close();
+      }
+    } catch (IOException e) {
+      logger.log(Level.WARNING, String.format("Failed to fetch event imagae from %s", imageUrl), e);
+      return new Response<>(500, "Interval Server Error", null);
+    }
+  }
+  
+  private AttachmentId getImageAttachmentId(String url) {
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java:339]: MAJOR Remove this unused private "getImageAttachmentId" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mwp/MwpClient.java [hunk 10]
```diff
@@ -41,7 +32,13 @@
   private Logger logger;
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mwp/MwpClient.java:34]: MINOR Remove the "logger" field and declare it as a local variable in the relevant methods. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1450)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/persistence/dao/OrganizationSettingDAO.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+   * @param value new value
+   * @return updated organization setting
+   */
+  public OrganizationSetting updateValue(OrganizationSetting setting, String value) {
+    setting.setValue(value);
+    return persist(setting);
+  }
+  
+  @Override
+  public void delete(OrganizationSetting setting) {
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/persistence/dao/OrganizationSettingDAO.java]: MINOR Remove this method to simply inherit it. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1185)

**Comment by sonarq** [comment]
SonarQube analysis reported 11 issues
* MAJOR 4 major
* MINOR 5 minor
* INFO 2 info

Watch the comments in this conversation to review them.

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+    result.setName(stripHtml(nytEvent.getName()));
+    result.setOriginalUrl(nytEvent.getUrl());
+    result.setPlace(stripHtml(nytEvent.getPlace()));
+    result.setZip(stripHtml(nytEvent.getZip()));
+    result.setStart(parseOffsetDateTime(nytEvent.getStart()));
+    
+    return result;
+  }
+  
+  private String stripHtml(String html) {
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java:392]: MAJOR Remove this unused private "stripHtml" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+    return StringUtils.trim(StringUtils.normalizeSpace(StringEscapeUtils.unescapeHtml4(html.replaceAll("\\<.*?>"," "))));
+  }
+  
+  private OffsetDateTime parseOffsetDateTime(String text) {
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/mikkelinyt/MikkeliNytEventProvider.java:396]: MAJOR Remove this unused private "parseOffsetDateTime" method. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

**Comment by sonarq** [comment]
SonarQube analysis reported 10 issues
* MAJOR 4 major
* MINOR 4 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by sonarq** [comment]
SonarQube analysis reported 8 issues
* MAJOR 4 major
* MINOR 2 minor
* INFO 2 info

Watch the comments in this conversation to review them.

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/EventProvider.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  
+  /**
+   * Returns data of event image
+   * 
+   * @param organizationId organization id
+   * @param eventId event id
+   * @param attachmentId image id
+   * @return event image data
+   */
+  public AttachmentData getEventImageData(OrganizationId organizationId, EventId eventId, AttachmentId attachmentId);
```
Comment (anttileppa) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/EventProvider.java]: add size parameter

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/GenericHttpCache.java [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import fi.otavanopisto.kuntaapi.server.integrations.GenericHttpClient.Response;
+import fi.otavanopisto.kuntaapi.server.integrations.GenericHttpClient.ResultType;
+
+/**
+ * Response aware client cache for integrations
+ * 
+ * @author Antti Leppä
```
Comment (anttileppa) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/integrations/GenericHttpCache.java:24]: remove stateless

File: kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/rest/OrganizationsApiImpl.java [hunk 16]
```diff
@@ -12,32 +17,50 @@
...
+ * 
+ * @author Antti Leppä
+ */
 @RequestScoped
 @Stateful
+@SuppressWarnings ("squid:S3306")
 public class OrganizationsApiImpl extends OrganizationsApi {
   
   @Inject
+  private Logger logger;
```
Comment (sonarq) [inline kunta-api-server/src/main/java/fi/otavanopisto/kuntaapi/server/rest/OrganizationsApiImpl.java:46]: MINOR Remove the "logger" field and declare it as a local variable in the relevant methods. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1450)

**Comment by sonarq** [comment]
SonarQube analysis reported 9 issues
- MAJOR 4 major
- MINOR 3 minor
- INFO 2 info

Watch the comments in this conversation to review them.

**Pull request merged**